### ISSUE TITLE
C++11 & result_of in odeint

### ISSUE THREAD

**Issue by ax3l** [ComputationalRadiationPhysics/picongpu#1704]
Using PIConGPU `dev` (C++11) as of 7bbd265e49a913a4cdc1c5d7d66ceeaa2e011314 with GCC 5.1, CUDA 7.5 & 8.0 and boost 1.58 ([module list](https://github.com/ComputationalRadiationPhysics/picongpu/blob/6e50572a1b778905aceee8914e2cf9021a916dd5/src/picongpu/submit/taurus-tud/picongpu.profile.example)) or 1.62 (C++11 self-build) on Taurus (TUD).

I am currently getting boost odeint (header only) compile time errors from the [synchrotron function integrator](https://github.com/ComputationalRadiationPhysics/picongpu/blob/6e50572a1b778905aceee8914e2cf9021a916dd5/src/picongpu/include/particles/synchrotronPhotons/SynchrotronFunctions.tpp#L77) of @Heikman in form of:

```
src/picongpu/include/particles/synchrotronPhotons/SynchrotronFunctions.tpp(77): here

/sw/global/compilers/gcc/5.1.0/include/c++/5.1.0/functional(268):
  error: class "boost::numeric::odeint::runge_kutta_dopri5<
    boost::array<picongpu::float_64, 1UL>,
    picongpu::float_64, boost::array<picongpu::float_64, 1UL>,
    picongpu::float_64, boost::numeric::odeint::array_algebra,
    boost::numeric::odeint::default_operations,
    boost::numeric::odeint::initially_resizer
  >" [...]

[...] has no member "result_type"
[...] has no member "argument_type"
[...] has no member "first_argument_type"
[...] has no member "second_argument_type"
```

I also tried using only `-DBOOST_RESULT_OF_USE_TR1` instead of the `-DBOOST_RESULT_OF_USE_TR1_WITH_DECLTYPE_FALLBACK` that is choosen right now but the error persists.

CCing @PrometheusPi @BeyondEspresso

**Comment by ax3l** [comment]
could be a problem that boost is using std `<functional>` instead of its own functionals.

**Comment by ax3l** [comment]
bug in HDF5 `1.8.15` header `H5api_adpt.h`:
```C
#ifdef __cplusplus
#define __attribute__(X)  /*void*/
#endif /* __cplusplus */
#ifndef H5_HAVE_ATTRIBUTE
#define __attribute__(X)  /*void*/
#endif /* H5_HAVE_ATTRIBUTE */
```

this is fixed in HDF5 `1.8.15-patch1`.

Those lines steals CUDA `__host__ __device__` instrumentations since they translate to an `__annotate__` which again translates to an `__attribute__` as found by @psychocoderHPC .

**Comment by ax3l** [comment]
the odeint bug is separate and can be solved by upgrading to `cuda/8.0.44` on Taurus.

**Comment by ax3l** [comment]
we don't investigate the boost issue deeper since we still have it covered by the compile suite and on hypnos and assume its something taurus specific with their "older" CUDA 7.5 install.
